File: src/utils/parseCard.ts
import { Card } from "../interfaces/card.interface";
import { rankMap } from "../types/rankMap";
import { suitMap } from "../types/suitMap";

/**
 * Parses a string like "AS", "TC", or "10D" into a Card object.
 */
export const parseCard = (cardStr: string): Card | null => {
  let rankChar: string;
  let suitChar: string;

  if (cardStr.length === 3 && cardStr.startsWith("10")) {
    rankChar = "T"; // Use 'T' to match the rankMap for '10'
    suitChar = cardStr[2].toUpperCase();
  } else if (cardStr.length === 2) {
    rankChar = cardStr[0].toUpperCase();
    suitChar = cardStr[1].toUpperCase();
  } else {
    // If not 2 or 3 characters, or doesn't start with "10", it's invalid
    console.error(`Error: Invalid card string format "${cardStr}"`);
    return null;
  }

  const rankInfo = rankMap[rankChar];
  const suit = suitMap[suitChar];

  if (!rankInfo || !suit) {
    console.error(`Error: Invalid rank or suit in card string "${cardStr}"`);
    return null;
  }

  return { ...rankInfo, suit };
};

File: crib_calc/utils/parse_card.py
from crib_calc.card import Card
from crib_calc.rank_map import RANK_MAP
from crib_calc.suit_map import SUIT_MAP


def parse_card(card_str: str) -> Card | None:
    if len(card_str) == 3 and card_str.startswith("10"):
        rank_char = "T"
        suit_char = card_str[2].upper()
    elif len(card_str) == 2:
        rank_char = card_str[0].upper()
        suit_char = card_str[1].upper()
    else:
        return None

    rank_info = RANK_MAP.get(rank_char)
    suit = SUIT_MAP.get(suit_char)

    if rank_info is None or suit is None:
        return None

    rank, value, order = rank_info
    return Card(suit=suit, rank=rank, value=value, order=order)
